### PR TITLE
Rails Authentication using devise gem

### PR DESCRIPTION
**Pull request by vikitaotiz**
The project tests the use of a devise gem to implement authentication.
I have created a posts controller and Post model.
I have protected the application controller using devise's before_action method

### CODE REVIEW

**Review by emmanuellekamwa — CHANGES_REQUESTED** [review]
**STATUS: CHANGES REQUESTED 🛠🛠**
Hello @vikitaotiz. Great job on your project so far but there are few more changes that you have to do to have this project approved.

- As a signed-in user, I should be able to create a new post that I can't. Please check this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_post_requirement.png) and kindly fix that.

- Also, as a signed-in user, I should see the authors of the different posts that I can't. Please check this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/signed_in_can_see_author_requirement.png) and kindly fix that.
- The posts routes creates the application routes using `
resources :posts, only: \[:new, :create, :index\]`
-  The Post controller uses `before_action` to restrict access to the `: new` and `: create` methods. Please don't forget to add that too at the top of your controller.

   Make the following changes and when you are ready request another code review. Good luck and happy coding 👨🏽‍💻👨🏽‍💻

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad and thanks for implementing the changes requested earlier on**. 

### Well-done 💯 
    1. The user can now create posts successfully.
    2. The linters are configured appropriately.

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [ ] Please take note that as requested previously, only a signed-in user can see the posts' author. When no user has signed in, the author of a post shouldn't be displayed.
- [ ] Check the inline comments for more sections to improve. 


#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+class PostsController < ApplicationController
+  skip_before_action :authenticate_user!, only: %i[index show]
```
Comment (mke2111) [inline app/controllers/posts_controller.rb]: - [ ] The Post controller uses `before_action` to restrict access to the `new` and `create` methods.

File: app/views/devise/mailer/email_changed.html.erb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
```
Comment (mke2111) [inline app/views/devise/mailer/email_changed.html.erb:3]: OPTIONAL

- [ ] It is a good practice to have zero logic in the views directory. Create methods in the helpers directory and call the methods in the views directory instead.
Read this [article](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) for a head-start. Another great piece to read is this [article](https://brandonhilkert.com/blog/refactoring-logic-from-a-rails-view/)

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Rails.application.routes.draw do
+  devise_for :users
+  root 'posts#index'
+  resources :posts
```
Comment (mke2111) [inline config/routes.rb]: - [ ] The posts routes creates the application routes using `resources :posts, only: \[:new, :create, :index\]`

**Review by mke2111 — CHANGES_REQUESTED** [review]
## Make Requested Changes 🔴 

### Please make the requested changes before requesting another CRR. If you need any form of assistance or more clarity about anything with regards to this project, reach out to @mke2111 or @emmanuellekamwa. We will be glad to help. 😄 

**Best regards.**

**Review by MkrtichSargsyan — CHANGES_REQUESTED** [review]
## _Status:_ Changes Requested :recycle:

## Hello Vikitaotiz Oluoch

THIS IS VERY IMPRESSIVE WORK. YOU HAVE DONE A GOOD JOB :+1:

## To Highlight :tada:

1. Your PR has a good title and description.
2. You have descriptive git commit messages.
3. Good job on implementing the changes requested.


However, there are some changes that you need to make in order to get your project approved


- [ ] The posts routes creates the application routes using resources :posts, only: \[:new, :create, :index\]

- [ ] Kindly use before_action to restrict access to the ```:new``` and ```:create``` methods in the Post controller 

- [ ] As a guest you can see only post's text [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/guest_can_see_only_posts_requirement.png)

<hr>


Please, make the REQUIRED changes and submit your request again! :grinning:
Good luck. 

> ps: If you have any questions about my review or need help, you can message me on slack ( @mkrtichsargsyan )

**Review by MkrtichSargsyan — APPROVED** [review]
## _Status:_ Approved  :hamster: :rocket:
Hello :wave:.

Awesome job on this project.:rocket: You have completed all the requested changes from the previous code reviewers.

You can go ahead and merge this branch to the base branch of this repo.
Good luck in your future projects and happy coding :relaxed:

> ps: . If you have any questions about my review or need help, you can message me on slack ( @mkrtichsargsyan )

**Pull request merged**